'raft applied is out of range' causing panic

I believe this is a recurence of #281

#### System information

Geth version: `geth version`
Geth
Version: 1.8.18-stable
Git Commit: c894c2d70eacf30740d03b53ed2fb39e42641295
Quorum Version: 2.2.5
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1337
Go Version: go1.11.12
Operating System: linux
GOPATH=
GOROOT=/usr/local/go
OS & Version: Windows/Linux/OSX

Branch, Commit Hash or Release: `git status`
Quorum Version: 2.2.5

#### Expected behaviour

After restart the node catches up with the chain

#### Actual behaviour

Node 1 partialy catches up, restarts and panics. there after it panics every restart
Node 2 partialy catches up, restarts and panics. there after it panics every restart
Node 1 & Node 2 reach the same block, but are about 200 blocks behind the minter at that time

#### Steps to reproduce the behaviour

* 5 node network
* Continually create tx's on the nodes (load testing at moderate rate)
* Take two nodes down for a few hours .

#### Backtrace

````
TRACE[10-13|08:57:13.777|eth/sync.go:89]                         Sending batch of transactions            id=c61a11c33dfeeba2 conn=staticdial count=10 bytes=25.43kB
INFO [10-13|08:57:14.177|raft/handler.go:468]                    startRaft                                raft ID=4
INFO [10-13|08:57:14.254|raft/handler.go:471]                    remounting an existing raft log; connecting to peers.
raft2019/10/13 08:57:14 applied(11688) is out of range [prevApplied(13718), committed(13718)]
panic: applied(11688) is out of range [prevApplied(13718), committed(13718)]

goroutine 1 [running]:
log.(*Logger).Panicf(0xc00013d9f0, 0x113ea16, 0x3c, 0xc003307260, 0x3, 0x3)
        /usr/local/go/src/log/log.go:219 +0xdb
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.(*DefaultLogger).Panicf(0x1f297f0, 0x113ea16, 0x3c, 0xc003307260, 0x3, 0x3)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/logger.go:121 +0x60
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.(*raftLog).appliedTo(0xc0003fe310, 0x2da8)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/log.go:202 +0x15b
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.newRaft(0xc003a73728, 0xc0001182a0)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/raft.go:306 +0xa97
github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft.RestartNode(0xc003a73728, 0x35, 0x0)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/github.com/coreos/etcd/raft/node.go:224 +0x40
github.com/ethereum/go-ethereum/raft.(*ProtocolManager).startRaft(0xc018d1a840)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/raft/handler.go:472 +0x7f4
github.com/ethereum/go-ethereum/raft.(*ProtocolManager).Start(0xc018d1a840, 0xc000427180)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/raft/handler.go:142 +0xd7
github.com/ethereum/go-ethereum/raft.(*RaftService).Start(0xc018f68f00, 0xc000427180, 0x0, 0x0)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/raft/backend.go:91 +0x39
github.com/ethereum/go-ethereum/node.(*Node).Start(0xc000426f00, 0x0, 0x0)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/node/node.go:206 +0xa92
github.com/ethereum/go-ethereum/cmd/utils.StartNode(0xc000426f00)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/utils/cmd.go:67 +0x2f
main.startNode(0xc00016c160, 0xc000426f00)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/geth/main.go:292 +0xb5
main.geth(0xc00016c160, 0x0, 0x0)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/geth/main.go:275 +0xe1
github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli%2ev1.HandleAction(0xf55700, 0x11af838, 0xc00016c160, 0xc00016c160, 0xc0004f9f58)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli.v1/app.go:490 +0xc8
github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli%2ev1.(*App).Run(0xc0001221a0, 0xc0000e4000, 0x23, 0x24, 0x0, 0x0)
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/vendor/gopkg.in/urfave/cli.v1/app.go:264 +0x59d
main.main()
        /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/geth/main.go:261 +0x55
````

#### Logs etc

We have a lot of log data from all nodes for this incident. If there is interest, I can work it into a useful shape.

--- a/raft/backend.go
+++ b/raft/backend.go
@@ -5,10 +5,9 @@ import (
 	"sync"
 	"time"
 
-	"github.com/ethereum/go-ethereum/core/types"
-
 	"github.com/ethereum/go-ethereum/accounts"
 	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/eth"
 	"github.com/ethereum/go-ethereum/eth/downloader"
 	"github.com/ethereum/go-ethereum/ethdb"

--- a/raft/handler.go
+++ b/raft/handler.go
@@ -426,6 +426,19 @@ func (pm *ProtocolManager) startRaft() {
 				// the single call to `pm.applyNewChainHead` for more context.
 				lastAppliedIndex = lastPersistedCommittedIndex
 			}
+
+			// fix raft applied index out of range
+			firstIndex, err := pm.raftStorage.FirstIndex()
+			if err != nil {
+				panic(fmt.Sprintf("failed to read last persisted applied index from raft while restarting: %v", err))
+			}
+			lastPersistedAppliedIndex := firstIndex - 1
+			if lastPersistedAppliedIndex > lastAppliedIndex {
+				log.Debug("set lastAppliedIndex to lastPersistedAppliedIndex", "last applied index", lastAppliedIndex, "last persisted applied index", lastPersistedAppliedIndex)
+
+				lastAppliedIndex = lastPersistedAppliedIndex
+				pm.advanceAppliedIndex(lastAppliedIndex)
+			}
 		}
 	}
 

--- a/raft/handler.go
+++ b/raft/handler.go
@@ -10,29 +10,27 @@ import (
 	"sync"
 	"time"
 
-	"golang.org/x/net/context"
-
+	"github.com/coreos/etcd/etcdserver/stats"
 	"github.com/coreos/etcd/pkg/fileutil"
+	raftTypes "github.com/coreos/etcd/pkg/types"
+	etcdRaft "github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/coreos/etcd/rafthttp"
 	"github.com/coreos/etcd/snap"
 	"github.com/coreos/etcd/wal"
+	mapset "github.com/deckarep/golang-set"
+	"github.com/syndtr/goleveldb/leveldb"
+	"golang.org/x/net/context"
+
 	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/eth/downloader"
 	"github.com/ethereum/go-ethereum/event"
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/p2p"
-	"github.com/ethereum/go-ethereum/rlp"
-
-	"github.com/coreos/etcd/etcdserver/stats"
-	raftTypes "github.com/coreos/etcd/pkg/types"
-	etcdRaft "github.com/coreos/etcd/raft"
-	"github.com/coreos/etcd/raft/raftpb"
-	"github.com/coreos/etcd/rafthttp"
 	"github.com/ethereum/go-ethereum/p2p/enode"
 	"github.com/ethereum/go-ethereum/p2p/enr"
-	"github.com/syndtr/goleveldb/leveldb"
-
-	mapset "github.com/deckarep/golang-set"
+	"github.com/ethereum/go-ethereum/rlp"
 )
 
 type ProtocolManager struct {

--- a/raft/handler_test.go
+++ b/raft/handler_test.go
@@ -1,0 +1,181 @@
+package raft
+
+import (
+	"crypto/ecdsa"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+	"unsafe"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum/go-ethereum/p2p/enr"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/ethash"
+	"github.com/ethereum/go-ethereum/core"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+
+	"github.com/ethereum/go-ethereum/eth"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// pm.advanceAppliedIndex() and state updates are in different
+// transaction boundaries hence there's a probablity that they are
+// out of sync due to premature shutdown
+func TestProtocolManager_whenAppliedIndexOutOfSync(t *testing.T) {
+	tmpWorkingDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_ = os.RemoveAll(tmpWorkingDir)
+	}()
+	count := 3
+	ports := make([]uint16, count)
+	nodeKeys := make([]*ecdsa.PrivateKey, count)
+	peers := make([]*enode.Node, count)
+	for i := 0; i < count; i++ {
+		ports[i] = nextPort(t)
+		nodeKeys[i] = mustNewNodeKey(t)
+		peers[i] = enode.NewV4(&(nodeKeys[i].PublicKey), net.IPv4(127, 0, 0, 1), 0, 0, int(ports[i]))
+	}
+	raftNodes := make([]*RaftService, count)
+	for i := 0; i < count; i++ {
+		if s, err := startRaftNode(uint16(i+1), ports[i], tmpWorkingDir, nodeKeys[i], peers); err != nil {
+			t.Fatal(err)
+		} else {
+			raftNodes[i] = s
+		}
+	}
+	waitFunc := func() {
+		for {
+			time.Sleep(200 * time.Millisecond)
+			for i := 0; i < count; i++ {
+				if raftNodes[i].raftProtocolManager.role == minterRole {
+					return
+				}
+			}
+		}
+	}
+	waitFunc()
+	// update the index to mimic the issue
+	raftNodes[0].raftProtocolManager.advanceAppliedIndex(1)
+	// now stop and restart the nodes
+	for i := 0; i < count; i++ {
+		if err := raftNodes[i].Stop(); err != nil {
+			t.Fatal(err)
+		}
+		for {
+			time.Sleep(200 * time.Millisecond)
+			if raftNodes[i].raftProtocolManager.stopped {
+				break
+			}
+		}
+	}
+	log.Debug("start raft cluster again")
+	for i := 0; i < count; i++ {
+		if s, err := startRaftNode(uint16(i+1), ports[i], tmpWorkingDir, nodeKeys[i], peers); err != nil {
+			t.Fatal(err)
+		} else {
+			raftNodes[i] = s
+		}
+	}
+	waitFunc()
+}
+
+func mustNewNodeKey(t *testing.T) *ecdsa.PrivateKey {
+	k, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	return k
+}
+
+func startRaftNode(id, port uint16, tmpWorkingDir string, key *ecdsa.PrivateKey, nodes []*enode.Node) (*RaftService, error) {
+	datadir := fmt.Sprintf("%s/node%d", tmpWorkingDir, id)
+	ctx, _, err := prepareServiceContext(key)
+	if err != nil {
+		return nil, err
+	}
+	ethCfg := &eth.Config{
+		Genesis:   &core.Genesis{Config: params.AllEthashProtocolChanges, GasLimit: 10000000000},
+		Etherbase: common.HexToAddress("0x8605cdbbdb6d264aa742e77020dcbc58fcdce182"),
+		Ethash: ethash.Config{
+			PowMode: ethash.ModeTest,
+		},
+	}
+	e, err := eth.New(ctx, ethCfg)
+	if err != nil {
+		return nil, err
+	}
+	s, err := New(ctx, params.QuorumTestChainConfig, id, port, false, 100*time.Millisecond, e, nodes, datadir)
+	if err != nil {
+		return nil, err
+	}
+	trustedID := enode.PubkeyToIDV4(&key.PublicKey)
+	srv := &p2p.Server{
+		Config: p2p.Config{
+			PrivateKey:   key,
+			MaxPeers:     10,
+			NoDial:       true,
+			TrustedNodes: []*enode.Node{newNode(trustedID, nil)},
+		},
+	}
+	if err := srv.Start(); err != nil {
+		return nil, fmt.Errorf("could not start: %v", err)
+	}
+	if err := s.Start(srv); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func newNode(id enode.ID, ip net.IP) *enode.Node {
+	var r enr.Record
+	if ip != nil {
+		r.Set(enr.IP(ip))
+	}
+	return enode.SignNull(&r, id)
+}
+
+func nextPort(t *testing.T) uint16 {
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	return uint16(listener.Addr().(*net.TCPAddr).Port)
+}
+
+func prepareServiceContext(key *ecdsa.PrivateKey) (ctx *node.ServiceContext, cfg *node.Config, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("%s", r)
+			ctx = nil
+			cfg = nil
+		}
+	}()
+	cfg = &node.Config{
+		P2P: p2p.Config{
+			PrivateKey: key,
+		},
+	}
+	ctx = &node.ServiceContext{
+		EventMux: new(event.TypeMux),
+	}
+	// config is private field so we need some workaround to set the value
+	configField := reflect.ValueOf(ctx).Elem().FieldByName("config")
+	configField = reflect.NewAt(configField.Type(), unsafe.Pointer(configField.UnsafeAddr())).Elem()
+	configField.Set(reflect.ValueOf(cfg))
+	return
+}

--- a/raft/handler_test.go
+++ b/raft/handler_test.go
@@ -11,22 +11,14 @@ import (
 	"time"
 	"unsafe"
 
-	"github.com/ethereum/go-ethereum/log"
-
-	"github.com/ethereum/go-ethereum/p2p/enr"
-
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/consensus/ethash"
 	"github.com/ethereum/go-ethereum/core"
-
-	"github.com/ethereum/go-ethereum/p2p/enode"
-
-	"github.com/ethereum/go-ethereum/eth"
-
 	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/eth"
 	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/node"
 	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
 	"github.com/ethereum/go-ethereum/params"
 )
 
@@ -69,21 +61,15 @@ func TestProtocolManager_whenAppliedIndexOutOfSync(t *testing.T) {
 		}
 	}
 	waitFunc()
-	// update the index to mimic the issue
+	// update the index to mimic the issue (set applied index behind for node 0)
 	raftNodes[0].raftProtocolManager.advanceAppliedIndex(1)
 	// now stop and restart the nodes
 	for i := 0; i < count; i++ {
 		if err := raftNodes[i].Stop(); err != nil {
 			t.Fatal(err)
 		}
-		for {
-			time.Sleep(200 * time.Millisecond)
-			if raftNodes[i].raftProtocolManager.stopped {
-				break
-			}
-		}
 	}
-	log.Debug("start raft cluster again")
+	log.Debug("restart raft cluster")
 	for i := 0; i < count; i++ {
 		if s, err := startRaftNode(uint16(i+1), ports[i], tmpWorkingDir, nodeKeys[i], peers); err != nil {
 			t.Fatal(err)
@@ -100,53 +86,6 @@ func mustNewNodeKey(t *testing.T) *ecdsa.PrivateKey {
 		t.Fatal(err)
 	}
 	return k
-}
-
-func startRaftNode(id, port uint16, tmpWorkingDir string, key *ecdsa.PrivateKey, nodes []*enode.Node) (*RaftService, error) {
-	datadir := fmt.Sprintf("%s/node%d", tmpWorkingDir, id)
-	ctx, _, err := prepareServiceContext(key)
-	if err != nil {
-		return nil, err
-	}
-	ethCfg := &eth.Config{
-		Genesis:   &core.Genesis{Config: params.AllEthashProtocolChanges, GasLimit: 10000000000},
-		Etherbase: common.HexToAddress("0x8605cdbbdb6d264aa742e77020dcbc58fcdce182"),
-		Ethash: ethash.Config{
-			PowMode: ethash.ModeTest,
-		},
-	}
-	e, err := eth.New(ctx, ethCfg)
-	if err != nil {
-		return nil, err
-	}
-	s, err := New(ctx, params.QuorumTestChainConfig, id, port, false, 100*time.Millisecond, e, nodes, datadir)
-	if err != nil {
-		return nil, err
-	}
-	trustedID := enode.PubkeyToIDV4(&key.PublicKey)
-	srv := &p2p.Server{
-		Config: p2p.Config{
-			PrivateKey:   key,
-			MaxPeers:     10,
-			NoDial:       true,
-			TrustedNodes: []*enode.Node{newNode(trustedID, nil)},
-		},
-	}
-	if err := srv.Start(); err != nil {
-		return nil, fmt.Errorf("could not start: %v", err)
-	}
-	if err := s.Start(srv); err != nil {
-		return nil, err
-	}
-	return s, nil
-}
-
-func newNode(id enode.ID, ip net.IP) *enode.Node {
-	var r enr.Record
-	if ip != nil {
-		r.Set(enr.IP(ip))
-	}
-	return enode.SignNull(&r, id)
 }
 
 func nextPort(t *testing.T) uint16 {
@@ -178,4 +117,39 @@ func prepareServiceContext(key *ecdsa.PrivateKey) (ctx *node.ServiceContext, cfg
 	configField = reflect.NewAt(configField.Type(), unsafe.Pointer(configField.UnsafeAddr())).Elem()
 	configField.Set(reflect.ValueOf(cfg))
 	return
+}
+
+func startRaftNode(id, port uint16, tmpWorkingDir string, key *ecdsa.PrivateKey, nodes []*enode.Node) (*RaftService, error) {
+	datadir := fmt.Sprintf("%s/node%d", tmpWorkingDir, id)
+
+	ctx, _, err := prepareServiceContext(key)
+	if err != nil {
+		return nil, err
+	}
+
+	e, err := eth.New(ctx, &eth.Config{
+		Genesis: &core.Genesis{Config: params.QuorumTestChainConfig},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := New(ctx, params.QuorumTestChainConfig, id, port, false, 100*time.Millisecond, e, nodes, datadir)
+	if err != nil {
+		return nil, err
+	}
+
+	srv := &p2p.Server{
+		Config: p2p.Config{
+			PrivateKey: key,
+		},
+	}
+	if err := srv.Start(); err != nil {
+		return nil, fmt.Errorf("could not start: %v", err)
+	}
+	if err := s.Start(srv); err != nil {
+		return nil, err
+	}
+
+	return s, nil
 }

--- a/raft/handler_test.go
+++ b/raft/handler_test.go
@@ -134,7 +134,7 @@ func startRaftNode(id, port uint16, tmpWorkingDir string, key *ecdsa.PrivateKey,
 		return nil, err
 	}
 
-	s, err := New(ctx, params.QuorumTestChainConfig, id, port, false, 100*time.Millisecond, e, nodes, datadir)
+	s, err := New(ctx, params.QuorumTestChainConfig, id, port, false, 100*time.Millisecond, e, nodes, datadir, false)
 	if err != nil {
 		return nil, err
 	}

--- a/raft/snapshot.go
+++ b/raft/snapshot.go
@@ -11,6 +11,7 @@ import (
 	"github.com/coreos/etcd/snap"
 	"github.com/coreos/etcd/wal/walpb"
 	mapset "github.com/deckarep/golang-set"
+
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/eth/downloader"

--- a/raft/snapshot.go
+++ b/raft/snapshot.go
@@ -10,7 +10,7 @@ import (
 	"github.com/coreos/etcd/raft/raftpb"
 	"github.com/coreos/etcd/snap"
 	"github.com/coreos/etcd/wal/walpb"
-	"github.com/deckarep/golang-set"
+	mapset "github.com/deckarep/golang-set"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/eth/downloader"

--- a/raft/speculative_chain.go
+++ b/raft/speculative_chain.go
@@ -1,11 +1,11 @@
 package raft
 
 import (
+	mapset "github.com/deckarep/golang-set"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/log"
 
-	"github.com/deckarep/golang-set"
 	"gopkg.in/oleiade/lane.v1"
 )
 

--- a/raft/speculative_chain.go
+++ b/raft/speculative_chain.go
@@ -2,11 +2,11 @@ package raft
 
 import (
 	mapset "github.com/deckarep/golang-set"
+	"gopkg.in/oleiade/lane.v1"
+
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/log"
-
-	"gopkg.in/oleiade/lane.v1"
 )
 
 // The speculative chain represents blocks that we have minted which haven't been accepted into the chain yet, building

